File: tests/java/com/pdfjet/MisuseTest.java
/*
 * MisuseTest.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
package com.pdfjet;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.pdfjet.encryption.Passwords;
import com.pdfjet.encryption.Permissions;
import java.io.ByteArrayOutputStream;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

/**
 * A program that uses the API the wrong way gets an exception, and complete()
 * then refuses to finish the document, so no broken PDF is written.
 */
class MisuseTest {
    private static final String EARLIER = "The PDF was not completed because of an earlier error: ";

    private static void assertRefused(PDF pdf, String message) {
        IllegalStateException e = assertThrows(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf.complete(); }
        });
        assertEquals(EARLIER + message, e.getMessage());
    }

    private static String fails(Class<? extends RuntimeException> type, Executable executable) {
        return assertThrows(type, executable).getMessage();
    }

    @Test
    void aNumberThatIsNotFiniteOrTooLargeIsRefused() throws Exception {
        for (final float bad : new float[] {Float.NaN, Float.POSITIVE_INFINITY, -1e30f, 2147483648f}) {
            final PDF pdf = TestSupport.newPDF();
            final Page page = new Page(pdf, Letter.PORTRAIT);
            String message = fails(IllegalArgumentException.class, new Executable() {
                public void execute() throws Throwable { page.drawLine(10f, 10f, bad, 200f); }
            });
            assertEquals("A coordinate, size or width is NaN, infinite or too large for a PDF.", message);
            assertRefused(pdf, message);
        }
        Page page = new Page(TestSupport.newPDF(), Letter.PORTRAIT);
        page.drawLine(0f, 0f, 100000000f, 0f);
        assertTrue(TestSupport.content(page).contains("100000000 792 l\n"), TestSupport.content(page));
    }

    @Test
    void aNaNFontSizeIsRefused() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        final Font font = TestSupport.helvetica(pdf);
        final Page page = new Page(pdf, Letter.PORTRAIT);
        font.setSize(Float.NaN);
        fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { new TextLine(font, "Hello").setLocation(50f, 50f).drawOn(page); }
        });
    }

    @Test
    void aDashPatternIsAnArrayAndAPhase() throws Exception {
        Page page = new Page(TestSupport.newPDF(), Letter.PORTRAIT);
        page.setStrokeDashPattern("[] 0");
        page.setStrokeDashPattern("[3 3] 0");
        page.setStrokeDashPattern(" [2.5 .5 1]  0.5 ");
        assertTrue(TestSupport.content(page).endsWith(" [2.5 .5 1]  0.5  d\n"), TestSupport.content(page));
        for (final String bad : new String[] {"3 3", "[3 3]", "[a] 0", "[0 0] 0", "[-1 2] 0", "[3-3] 0", "[1.5.5] 0", "[3 3] 0 ET", ""}) {
            final PDF pdf = TestSupport.newPDF();
            final Page page2 = new Page(pdf, Letter.PORTRAIT);
            String message = fails(IllegalArgumentException.class, new Executable() {
                public void execute() throws Throwable { page2.setStrokeDashPattern(bad); }
            });
            assertEquals("The dash pattern \"" + bad + "\" is not an array of non-negative numbers, "
                    + "not all zero, followed by a phase, such as \"[3 3] 0\".", message);
            assertEquals("", TestSupport.content(page2));
        }
    }

    @Test
    void aNegativePenWidthIsRefused() throws Exception {
        final Page page = new Page(TestSupport.newPDF(), Letter.PORTRAIT);
        assertEquals("The pen width cannot be negative.", fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { page.setPenWidth(-2f); }
        }));
        page.setPenWidth(0f);
    }

    @Test
    void theGraphicsStatesMustBePaired() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        final Page page = new Page(pdf, Letter.PORTRAIT);
        String message = fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { page.restoreGraphicsState(); }
        });
        assertEquals("restoreGraphicsState was called without a matching saveGraphicsState.", message);
        assertEquals("", TestSupport.content(page));

        final PDF pdf2 = TestSupport.newPDF();
        new Page(pdf2, Letter.PORTRAIT).saveGraphicsState();
        message = fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { new Page(pdf2, Letter.PORTRAIT); }
        });
        assertEquals("A page ends with a saveGraphicsState that has no restoreGraphicsState.", message);

        final PDF pdf3 = TestSupport.newPDF();
        new Page(pdf3, Letter.PORTRAIT).saveGraphicsState();
        assertEquals("A page ends with a saveGraphicsState that has no restoreGraphicsState.",
                fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf3.complete(); }
        }));
    }

    @Test
    void theMarkedContentMustBePaired() throws Exception {
        for (Compliance compliance : new Compliance[] {Compliance.PDF_1_7, Compliance.PDF_UA_1}) {
            PDF pdf = new PDF(new ByteArrayOutputStream(), compliance);
            final Page page = new Page(pdf, Letter.PORTRAIT);
            String message = fails(IllegalStateException.class, new Executable() {
                public void execute() throws Throwable { page.addEMC(); }
            });
            assertEquals("addEMC was called without a matching addBDC or addArtifactBMC.", message);

            final PDF pdf2 = new PDF(new ByteArrayOutputStream(), compliance);
            new Page(pdf2, Letter.PORTRAIT).addBDC(StructElem.P, "x", "x");
            assertEquals("A page ends with an addBDC or addArtifactBMC that has no addEMC.",
                    fails(IllegalStateException.class, new Executable() {
                public void execute() throws Throwable { pdf2.complete(); }
            }));

            PDF pdf3 = new PDF(new ByteArrayOutputStream(), compliance);
            Page page3 = new Page(pdf3, Letter.PORTRAIT);
            page3.addArtifactBMC();
            page3.addEMC();
            pdf3.complete();
        }
    }

    @Test
    void aWrittenPageCannotBeDrawnOn() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        final Font font = TestSupport.helvetica(pdf);
        final Page page1 = new Page(pdf, Letter.PORTRAIT);
        new Page(pdf, Letter.PORTRAIT);
        String message = fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { new TextLine(font, "Late").setLocation(50f, 50f).drawOn(page1); }
        });
        assertEquals("The page was already written to the PDF: draw on a page before "
                + "creating the next page or completing the PDF.", message);
        assertRefused(pdf, message);
    }

    @Test
    void completeFinishesADocumentOnce() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        new Page(pdf, Letter.PORTRAIT);
        pdf.complete();
        assertEquals("complete() was already called.", fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf.complete(); }
        }));
        assertEquals("The PDF was already completed.", fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { new Page(pdf, Letter.PORTRAIT); }
        }));
    }

    @Test
    void aDocumentNeedsAPage() throws Exception {
        PDF pdf = TestSupport.newPDF();
        IllegalStateException e = assertThrows(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf.complete(); }
        });
        assertEquals("A PDF needs at least one page.", e.getMessage());
    }

    @Test
    void aPageIsAddedOnceAndToItsOwnDocument() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        final Page page = new Page(pdf, Letter.PORTRAIT, Page.DETACHED);
        pdf.addPage(page);
        assertEquals("The page was already added to the PDF.", fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf.addPage(page); }
        }));

        final PDF pdf2 = TestSupport.newPDF();
        final Page foreign = new Page(TestSupport.newPDF(), Letter.PORTRAIT, Page.DETACHED);
        assertEquals("The page belongs to another PDF.", fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { pdf2.addPage(foreign); }
        }));
    }

    @Test
    void fontsImagesStampsAndGroupsBelongToOneDocument() throws Exception {
        final PDF other = TestSupport.newPDF();
        final PDF pdf = TestSupport.newPDF();
        final Page page = new Page(pdf, Letter.PORTRAIT);

        final Font font = TestSupport.helvetica(other);
        assertEquals("The font belongs to another PDF.", fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { new TextLine(font, "Hello").setLocation(50f, 50f).drawOn(page); }
        }));

        final Image image = new Image(other, TestSupport.open("images/map407.png"));
        assertEquals("The image belongs to another PDF.", fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { image.setLocation(50f, 50f).drawOn(page); }
        }));

        final Stamp stamp = new Stamp(other).setSize(50f, 50f);
        stamp.complete();
        assertEquals("The stamp belongs to another PDF.", fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { stamp.drawOn(page); }
        }));
        final Stamp stamp2 = new Stamp(pdf).setSize(50f, 50f);
        assertEquals("The font belongs to another PDF.", fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { stamp2.addFont(font); }
        }));

        final OptionalContentGroup group = new OptionalContentGroup(other, "Layer");
        group.add(new Rect(0f, 0f, 1f, 1f));
        assertEquals("The optional content group belongs to another PDF.",
                fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { group.drawOn(page); }
        }));
    }

    @Test
    void aStampIsCompletedOnceBeforeItIsDrawn() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        final Page page = new Page(pdf, Letter.PORTRAIT);
        final Stamp stamp = new Stamp(pdf).setSize(50f, 50f);
        assertEquals("Call complete() on the stamp before drawing it.",
                fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { stamp.drawOn(page); }
        }));

        final PDF pdf2 = TestSupport.newPDF();
        final Stamp stamp2 = new Stamp(pdf2).setSize(50f, 50f);
        stamp2.complete();
        assertEquals("complete() was already called on the stamp.", fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { stamp2.complete(); }
        }));
        assertEquals("The stamp was already completed.", fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { stamp2.drawRect(0f, 0f, 10f, 10f); }
        }));
    }

    @Test
    void stampTextNeedsAFontAndAText() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        final Stamp stamp = new Stamp(pdf).setSize(100f, 50f);
        for (final TextParameters parameters : new TextParameters[] {
                new TextParameters().setText("Paid"), new TextParameters().setFont(TestSupport.helvetica(pdf))}) {
            assertEquals("Stamp text needs a font and a text.", fails(IllegalArgumentException.class, new Executable() {
                public void execute() throws Throwable { stamp.drawText(parameters); }
            }));
        }
    }

    @Test
    void stampTextAddsItsFont() throws Exception {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        final PDF pdf = new PDF(bos);
        Font font = new Font(pdf, TestSupport.open("fonts/IBMPlexSans/IBMPlexSans-Regular.otf.stream"));
        final Font core = TestSupport.helvetica(pdf);
        final Stamp stamp = new Stamp(pdf).setSize(100f, 50f);
        assertEquals("A stamp draws text with an embedded font, not a core or CJK font.",
                fails(IllegalArgumentException.class, new Executable() {
            public void execute() throws Throwable { stamp.drawText(core, 12f, 5f, 20f, "Paid"); }
        }));
        stamp.drawText(font, 12f, 5f, 20f, "Paid");
        stamp.complete();
        String raw = TestSupport.latin1(bos.toByteArray());
        assertTrue(raw.contains("/Font <<\n/F" + font.objNumber + " " + font.objNumber + " 0 R\n"), raw);
    }

    @Test
    void encryptionAndComplianceComeBeforeTheContent() throws Exception {
        final PDF pdf = TestSupport.newPDF();
        TestSupport.helvetica(pdf);
        assertEquals("Set the encryption before adding fonts, images or pages to the PDF.",
                fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { new Encryption(pdf, new Passwords(), new Permissions()); }
        }));
        assertEquals("Set the compliance before adding fonts, images or pages to the PDF.",
                fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf.setCompliance(Compliance.PDF_UA_1); }
        }));
        pdf.setCompliance(Compliance.PDF_1_7);   // No change.

        final PDF pdf2 = TestSupport.newPDF();
        new Page(pdf2, Letter.PORTRAIT, Page.DETACHED);
        assertEquals("Set the compliance before adding fonts, images or pages to the PDF.",
                fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf2.setCompliance(Compliance.PDF_UA_1); }
        }));

        final PDF pdf3 = TestSupport.newPDF();
        final Encryption encryption = new Encryption(pdf3, new Passwords(), new Permissions());
        TestSupport.helvetica(pdf3);
        assertEquals("Set the encryption before adding fonts, images or pages to the PDF.",
                fails(IllegalStateException.class, new Executable() {
            public void execute() throws Throwable { pdf3.setEncryption(encryption); }
        }));
    }

    @Test
    void aPageIsFromThreeTo14400PointsWideAndHigh() throws Exception {
        new Page(TestSupport.newPDF(), new PageSize(3f, 3f));
        new Page(TestSupport.newPDF(), new PageSize(14400f, 14400f));
        for (final PageSize bad : new PageSize[] {
                new PageSize(0f, 0f), new PageSize(612f, 2f), new PageSize(14401f, 792f), new PageSize(Float.NaN, 792f)}) {
            final PDF pdf = TestSupport.newPDF();
            assertEquals("A page must be from 3 to 14400 points wide and high.",
                    fails(IllegalArgumentException.class, new Executable() {
                public void execute() throws Throwable { new Page(pdf, bad); }
            }));
        }
    }

    @Test
    void theXmpMetadataLeavesOutControlCharacters() throws Exception {
        ByteArrayOutputStream bos = new ByteArrayOutputStream();
        PDF pdf = new PDF(bos, Compliance.PDF_UA_1);
        pdf.setTitle("Report\u0001 2026\uFFFF \uD800");
        new Page(pdf, Letter.PORTRAIT);
        pdf.complete();
        String raw = new String(bos.toByteArray(), "UTF-8");
        assertTrue(raw.contains("<rdf:li xml:lang=\"x-default\">Report 2026 </rdf:li>"), raw);
    }

    @Test
    void aZeroSizeImageStampOrContainerDrawsNothing() throws Exception {
        PDF pdf = TestSupport.newPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        Image image = new Image(pdf, TestSupport.open("images/map407.png"));
        image.scaleBy(0f);
        image.setLocation(50f, 50f).drawOn(page);
        Stamp stamp = new Stamp(pdf).setSize(50f, 50f);
        stamp.complete();
        stamp.scaleBy(0f).setLocation(50f, 50f).drawOn(page);
        Container container = new Container(100f, 100f);
        container.add(new Rect(0f, 0f, 10f, 10f));
        container.scaleBy(0f).setLocation(50f, 50f).drawOn(page);
        assertFalse(TestSupport.content(page).contains(" cm\n"), TestSupport.content(page));
    }
}

File: tests/dotnet/MisuseTest.cs
/*
 * MisuseTest.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;
using System.IO;
using System.Text;
using System.Collections.Generic;
using Xunit;

namespace PDFjet.NET {
/// <summary>
/// A program that uses the API the wrong way gets an exception, and Complete()
/// then refuses to finish the document, so no broken PDF is written.
/// </summary>
public class MisuseTest {
    private const string EARLIER = "The PDF was not completed because of an earlier error: ";

    private static void AssertRefused(PDF pdf, string message) {
        InvalidOperationException e = Assert.Throws<InvalidOperationException>(() => pdf.Complete());
        Assert.Equal(EARLIER + message, e.Message);
    }

    [Fact]
    public void ANumberThatIsNotFiniteOrTooLargeIsRefused() {
        foreach (float bad in new float[] {float.NaN, float.PositiveInfinity, -1e30f, 2147483648f}) {
            PDF pdf = TestSupport.NewPDF();
            Page page = new Page(pdf, Letter.PORTRAIT);
            string message = Assert.Throws<ArgumentException>(() => page.DrawLine(10f, 10f, bad, 200f)).Message;
            Assert.Equal("A coordinate, size or width is NaN, infinite or too large for a PDF.", message);
            AssertRefused(pdf, message);
        }
        Page page2 = new Page(TestSupport.NewPDF(), Letter.PORTRAIT);
        page2.DrawLine(0f, 0f, 100000000f, 0f);
        Assert.Contains("100000000 792 l\n", TestSupport.Content(page2));
    }

    [Fact]
    public void ANaNFontSizeIsRefused() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page = new Page(pdf, Letter.PORTRAIT);
        font.SetSize(float.NaN);
        Assert.Throws<ArgumentException>(() => new TextLine(font, "Hello").SetLocation(50f, 50f).DrawOn(page));
    }

    [Fact]
    public void ADashPatternIsAnArrayAndAPhase() {
        Page page = new Page(TestSupport.NewPDF(), Letter.PORTRAIT);
        page.SetStrokeDashPattern("[] 0");
        page.SetStrokeDashPattern("[3 3] 0");
        page.SetStrokeDashPattern(" [2.5 .5 1]  0.5 ");
        Assert.EndsWith(" [2.5 .5 1]  0.5  d\n", TestSupport.Content(page));
        foreach (string bad in new string[] {"3 3", "[3 3]", "[a] 0", "[0 0] 0", "[-1 2] 0", "[3-3] 0", "[1.5.5] 0", "[3 3] 0 ET", ""}) {
            PDF pdf = TestSupport.NewPDF();
            Page page2 = new Page(pdf, Letter.PORTRAIT);
            string message = Assert.Throws<ArgumentException>(() => page2.SetStrokeDashPattern(bad)).Message;
            Assert.Equal("The dash pattern \"" + bad + "\" is not an array of non-negative numbers, "
                    + "not all zero, followed by a phase, such as \"[3 3] 0\".", message);
            Assert.Equal("", TestSupport.Content(page2));
        }
    }

    [Fact]
    public void ANegativePenWidthIsRefused() {
        Page page = new Page(TestSupport.NewPDF(), Letter.PORTRAIT);
        Assert.Equal("The pen width cannot be negative.",
                Assert.Throws<ArgumentException>(() => page.SetPenWidth(-2f)).Message);
        page.SetPenWidth(0f);
    }

    [Fact]
    public void TheGraphicsStatesMustBePaired() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        string message = Assert.Throws<InvalidOperationException>(() => page.RestoreGraphicsState()).Message;
        Assert.Equal("RestoreGraphicsState was called without a matching SaveGraphicsState.", message);
        Assert.Equal("", TestSupport.Content(page));

        PDF pdf2 = TestSupport.NewPDF();
        new Page(pdf2, Letter.PORTRAIT).SaveGraphicsState();
        message = Assert.Throws<InvalidOperationException>(() => new Page(pdf2, Letter.PORTRAIT)).Message;
        Assert.Equal("A page ends with a SaveGraphicsState that has no RestoreGraphicsState.", message);

        PDF pdf3 = TestSupport.NewPDF();
        new Page(pdf3, Letter.PORTRAIT).SaveGraphicsState();
        Assert.Equal("A page ends with a SaveGraphicsState that has no RestoreGraphicsState.",
                Assert.Throws<InvalidOperationException>(() => pdf3.Complete()).Message);
    }

    [Fact]
    public void TheMarkedContentMustBePaired() {
        foreach (Compliance compliance in new Compliance[] {Compliance.PDF_1_7, Compliance.PDF_UA_1}) {
            PDF pdf = new PDF(new MemoryStream(), compliance);
            Page page = new Page(pdf, Letter.PORTRAIT);
            string message = Assert.Throws<InvalidOperationException>(() => page.AddEMC()).Message;
            Assert.Equal("AddEMC was called without a matching AddBDC or AddArtifactBMC.", message);

            PDF pdf2 = new PDF(new MemoryStream(), compliance);
            new Page(pdf2, Letter.PORTRAIT).AddBDC(StructElem.P, "x", "x");
            Assert.Equal("A page ends with an AddBDC or AddArtifactBMC that has no AddEMC.",
                    Assert.Throws<InvalidOperationException>(() => pdf2.Complete()).Message);

            PDF pdf3 = new PDF(new MemoryStream(), compliance);
            Page page3 = new Page(pdf3, Letter.PORTRAIT);
            page3.AddArtifactBMC();
            page3.AddEMC();
            pdf3.Complete();
        }
    }

    [Fact]
    public void AWrittenPageCannotBeDrawnOn() {
        PDF pdf = TestSupport.NewPDF();
        Font font = TestSupport.Helvetica(pdf);
        Page page1 = new Page(pdf, Letter.PORTRAIT);
        new Page(pdf, Letter.PORTRAIT);
        string message = Assert.Throws<InvalidOperationException>(
                () => new TextLine(font, "Late").SetLocation(50f, 50f).DrawOn(page1)).Message;
        Assert.Equal("The page was already written to the PDF: draw on a page before "
                + "creating the next page or completing the PDF.", message);
        AssertRefused(pdf, message);
    }

    [Fact]
    public void CompleteFinishesADocumentOnce() {
        PDF pdf = TestSupport.NewPDF();
        new Page(pdf, Letter.PORTRAIT);
        pdf.Complete();
        Assert.Equal("Complete() was already called.",
                Assert.Throws<InvalidOperationException>(() => pdf.Complete()).Message);
        Assert.Equal("The PDF was already completed.",
                Assert.Throws<InvalidOperationException>(() => new Page(pdf, Letter.PORTRAIT)).Message);
    }

    [Fact]
    public void ADocumentNeedsAPage() {
        PDF pdf = TestSupport.NewPDF();
        Assert.Equal("A PDF needs at least one page.",
                Assert.Throws<InvalidOperationException>(() => pdf.Complete()).Message);
    }

    [Fact]
    public void APageIsAddedOnceAndToItsOwnDocument() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT, Page.DETACHED);
        pdf.AddPage(page);
        Assert.Equal("The page was already added to the PDF.",
                Assert.Throws<InvalidOperationException>(() => pdf.AddPage(page)).Message);

        PDF pdf2 = TestSupport.NewPDF();
        Page foreign = new Page(TestSupport.NewPDF(), Letter.PORTRAIT, Page.DETACHED);
        Assert.Equal("The page belongs to another PDF.",
                Assert.Throws<ArgumentException>(() => pdf2.AddPage(foreign)).Message);
    }

    [Fact]
    public void FontsImagesStampsAndGroupsBelongToOneDocument() {
        PDF other = TestSupport.NewPDF();
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);

        Font font = TestSupport.Helvetica(other);
        Assert.Equal("The font belongs to another PDF.", Assert.Throws<ArgumentException>(
                () => new TextLine(font, "Hello").SetLocation(50f, 50f).DrawOn(page)).Message);

        Image image = new Image(other, TestSupport.Open("PngSuite/BASN2C08.PNG"));
        Assert.Equal("The image belongs to another PDF.", Assert.Throws<ArgumentException>(
                () => image.SetLocation(50f, 50f).DrawOn(page)).Message);

        Stamp stamp = new Stamp(other).SetSize(50f, 50f);
        stamp.Complete();
        Assert.Equal("The stamp belongs to another PDF.",
                Assert.Throws<ArgumentException>(() => stamp.DrawOn(page)).Message);
        Stamp stamp2 = new Stamp(pdf).SetSize(50f, 50f);
        Assert.Equal("The font belongs to another PDF.",
                Assert.Throws<ArgumentException>(() => stamp2.AddFont(font)).Message);

        OptionalContentGroup group = new OptionalContentGroup(other, "Layer");
        group.Add(new Rect(0f, 0f, 1f, 1f));
        Assert.Equal("The optional content group belongs to another PDF.",
                Assert.Throws<ArgumentException>(() => group.DrawOn(page)).Message);
    }

    [Fact]
    public void StampTextNeedsAFontAndAText() {
        PDF pdf = TestSupport.NewPDF();
        Stamp stamp = new Stamp(pdf).SetSize(100f, 50f);
        foreach (TextParameters parameters in new TextParameters[] {
                new TextParameters().SetText("Paid"), new TextParameters().SetFont(TestSupport.Helvetica(pdf))}) {
            Assert.Equal("Stamp text needs a font and a text.",
                    Assert.Throws<ArgumentException>(() => stamp.DrawText(parameters)).Message);
        }
    }

    [Fact]
    public void AStampIsCompletedOnceBeforeItIsDrawn() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        Stamp stamp = new Stamp(pdf).SetSize(50f, 50f);
        Assert.Equal("Call Complete() on the stamp before drawing it.",
                Assert.Throws<InvalidOperationException>(() => stamp.DrawOn(page)).Message);

        PDF pdf2 = TestSupport.NewPDF();
        Stamp stamp2 = new Stamp(pdf2).SetSize(50f, 50f);
        stamp2.Complete();
        Assert.Equal("Complete() was already called on the stamp.",
                Assert.Throws<InvalidOperationException>(() => stamp2.Complete()).Message);
        Assert.Equal("The stamp was already completed.",
                Assert.Throws<InvalidOperationException>(() => stamp2.DrawRect(0f, 0f, 10f, 10f)).Message);
    }

    [Fact]
    public void StampTextAddsItsFont() {
        MemoryStream stream = new MemoryStream();
        PDF pdf = new PDF(stream);
        Font font = new Font(pdf, TestSupport.Open("fonts/IBMPlexSans/IBMPlexSans-Regular.otf.stream"));
        Font core = TestSupport.Helvetica(pdf);
        Stamp stamp = new Stamp(pdf).SetSize(100f, 50f);
        Assert.Equal("A stamp draws text with an embedded font, not a core or CJK font.",
                Assert.Throws<ArgumentException>(() => stamp.DrawText(core, 12f, 5f, 20f, "Paid")).Message);
        stamp.DrawText(font, 12f, 5f, 20f, "Paid");
        stamp.Complete();
        string raw = TestSupport.Latin1(stream.ToArray());
        Assert.Contains("/Font <<\n/F" + font.objNumber + " " + font.objNumber + " 0 R\n", raw);
    }

    [Fact]
    public void EncryptionAndComplianceComeBeforeTheContent() {
        PDF pdf = TestSupport.NewPDF();
        TestSupport.Helvetica(pdf);
        Assert.Equal("Set the encryption before adding fonts, images or pages to the PDF.",
                Assert.Throws<InvalidOperationException>(
                        () => new Encryption(pdf, new Passwords(), new Permissions())).Message);
        Assert.Equal("Set the compliance before adding fonts, images or pages to the PDF.",
                Assert.Throws<InvalidOperationException>(() => pdf.SetCompliance(Compliance.PDF_UA_1)).Message);
        pdf.SetCompliance(Compliance.PDF_1_7);   // No change.

        PDF pdf2 = TestSupport.NewPDF();
        new Page(pdf2, Letter.PORTRAIT, Page.DETACHED);
        Assert.Equal("Set the compliance before adding fonts, images or pages to the PDF.",
                Assert.Throws<InvalidOperationException>(() => pdf2.SetCompliance(Compliance.PDF_UA_1)).Message);

        PDF pdf3 = TestSupport.NewPDF();
        Encryption encryption = new Encryption(pdf3, new Passwords(), new Permissions());
        TestSupport.Helvetica(pdf3);
        Assert.Equal("Set the encryption before adding fonts, images or pages to the PDF.",
                Assert.Throws<InvalidOperationException>(() => pdf3.SetEncryption(encryption)).Message);
    }

    [Fact]
    public void APageIsFromThreeTo14400PointsWideAndHigh() {
        new Page(TestSupport.NewPDF(), new PageSize(3f, 3f));
        new Page(TestSupport.NewPDF(), new PageSize(14400f, 14400f));
        foreach (PageSize bad in new PageSize[] {
                new PageSize(0f, 0f), new PageSize(612f, 2f), new PageSize(14401f, 792f), new PageSize(float.NaN, 792f)}) {
            PDF pdf = TestSupport.NewPDF();
            Assert.Equal("A page must be from 3 to 14400 points wide and high.",
                    Assert.Throws<ArgumentException>(() => new Page(pdf, bad)).Message);
        }
    }

    [Fact]
    public void TheXmpMetadataLeavesOutControlCharacters() {
        MemoryStream stream = new MemoryStream();
        PDF pdf = new PDF(stream, Compliance.PDF_UA_1);
        pdf.SetTitle("Report\u0001 2026\uFFFF \uD800");
        new Page(pdf, Letter.PORTRAIT);
        pdf.Complete();
        string raw = Encoding.UTF8.GetString(stream.ToArray());
        Assert.Contains("<rdf:li xml:lang=\"x-default\">Report 2026 </rdf:li>", raw);
    }

    [Fact]
    public void AZeroSizeImageStampOrContainerDrawsNothing() {
        PDF pdf = TestSupport.NewPDF();
        Page page = new Page(pdf, Letter.PORTRAIT);
        Image image = new Image(pdf, TestSupport.Open("PngSuite/BASN2C08.PNG"));
        image.ScaleBy(0f);
        image.SetLocation(50f, 50f).DrawOn(page);
        Stamp stamp = new Stamp(pdf).SetSize(50f, 50f);
        stamp.Complete();
        stamp.ScaleBy(0f).SetLocation(50f, 50f).DrawOn(page);
        Container container = new Container(100f, 100f);
        container.Add(new Rect(0f, 0f, 10f, 10f));
        container.ScaleBy(0f).SetLocation(50f, 50f).DrawOn(page);
        Assert.DoesNotContain(" cm\n", TestSupport.Content(page));
    }

    // The objects of a small PDF, as Read returns them.
    private static List<PDFobj> ExistingObjects() {
        MemoryStream stream = new MemoryStream();
        PDF pdf = new PDF(stream);
        new TextLine(TestSupport.Helvetica(pdf), "Existing").SetLocation(50f, 50f).DrawOn(new Page(pdf, Letter.PORTRAIT));
        pdf.Complete();
        return TestSupport.Read(stream.ToArray());
    }

    [Fact]
    public void TheObjectsOfAnExistingPdfComeBeforeTheContent() {
        List<PDFobj> objects = ExistingObjects();
        string message = "Add the objects of an existing PDF before fonts, images or pages "
                + "are added to the PDF: object 1 is already written.";

        PDF pdf = TestSupport.NewPDF();
        TestSupport.Helvetica(pdf);     // Object 1, which the objects would replace.
        Assert.Equal(message,
                Assert.Throws<InvalidOperationException>(() => pdf.AddResourceObjects(objects)).Message);
        AssertRefused(pdf, message);

        PDF pdf2 = TestSupport.NewPDF();
        TestSupport.Helvetica(pdf2);
        Assert.Equal(message,
                Assert.Throws<InvalidOperationException>(() => pdf2.AddObjects(objects)).Message);

        // The objects first, and the font after them.
        PDF pdf3 = TestSupport.NewPDF();
        pdf3.AddResourceObjects(objects);
        TestSupport.Helvetica(pdf3);
        new Page(pdf3, Letter.PORTRAIT);
        pdf3.Complete();
    }

    [Fact]
    public void TheObjectsOfAnExistingPdfCannotBeAddedToAnEncryptedPdf() {
        List<PDFobj> objects = ExistingObjects();
        PDF pdf = TestSupport.NewPDF();
        pdf.SetEncryption(new Encryption(pdf, new Passwords(), new Permissions()));
        string message = "The objects of an existing PDF cannot be added to an encrypted PDF.";
        Assert.Equal(message,
                Assert.Throws<InvalidOperationException>(() => pdf.AddResourceObjects(objects)).Message);
        AssertRefused(pdf, message);
    }

    [Fact]
    public void APdfACannotBeEncrypted() {
        PDF pdf = new PDF(new MemoryStream(), Compliance.PDF_A_2B);
        Encryption encryption = new Encryption(pdf, new Passwords(), new Permissions());
        string message = "A PDF/A document cannot be encrypted.";
        Assert.Equal(message,
                Assert.Throws<InvalidOperationException>(() => pdf.SetEncryption(encryption)).Message);
        AssertRefused(pdf, message);

        // A PDF/UA document can be.
        PDF ua = new PDF(new MemoryStream(), Compliance.PDF_UA_1);
        ua.SetEncryption(new Encryption(ua, new Passwords(), new Permissions()));
    }
}
}
